keps/sig-scheduling/4247-queueinghint/README.md

```diff
@@ -0,0 +1,1122 @@
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+- If a plugin' QueueingHint implementation has bugs and, for example, misses some events that can make Pods schedulable,
+Pods rejected by those plugins may be stuck in the unschedulable Pod pool for a long time.
+  - Detection: Pods get `FailedScheduling` event, but not retried during 5 min even if the cluster should have a state that can accommodate those Pods. 
+  - Mitigations: The scheduling queue priodically flushing Pods in the unschedulable Pod pool. So, even if such bug exists, Pods' scheduling are retried after a certain period, which is 5 min by default. You can shorten the max duration that Pods can stay in the unschedulable Pod pool by using `--pod-max-in-unschedulable-pods-duration`.
+  - Diagnostics: If you increases the log level to more than 5, you can see the logs related to `QueueingHint` in the scheduling queue. Also, the in-tree plugins emits all logs in QueueingHint with log level 5. (If you have a custom plugin, you may want to check the log level in its QueueingHint.)
+  - Testing: There are multiple unit tests to confirm `flushUnschedulablePodsLeftover` is working expectedly.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+- Jun 26, 2023: The QueueingHint is implemented and the `EnqueueExtension` interface is changed.
+- Jul 15, 2023: The feature gate is implemented. (enabled by default)
+- Jul 18, 2023: The scheduling queue tracks the Pod being processed to put it back to an appropriate queue.
+- Oct 01, 2023: The initial KEP is submitted.
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Return `QueueImmediately`, `QueueAfterBackoff`, and `QueueSkip` from `QueueingHintFn` instead of introducing new status `Pending`
```

Comment:
Personally, I prefer this one because it's more simple and extensible, people can decide the requeueing policy by the hint only, decouple with framework status, e.g. when implementing custom plugins, they can still return unschedulable status but return `QueueImmediately` in hintFunc if necessary. Also it's backwards-compatible.

If we want to use `Pending` status, can `Wait` fit for this, then no need to introduce a new status.

Reply:
> they can still return unschedulable status but return QueueImmediately in hintFunc if necessary.

So, when could they want to return QueueImmediately but need to return unschedulable in the scheduling cycle?
More flexibility/extensibility is not always better. All plugins (including custom ones) gather in the scheduler and work together. So, we need a clear rule and guardrail to prevent incorrect usage of when to skip backoff and when to not skip. Otherwise, everyone has a different idea like we saw in NodeAffinity QueueingHint PR and each plugin may be implemented in different strategies. 
In the current KEP, if we define the beckoff delay as a penalty of wasting the scheduling cycle, to define how to requeue Pods (go thru backoff or not) with the returned status makes more sense.

If we want to go away from this design and keep using `QueueImmediately`, `QueueAfterBackoff`, and `QueueSkip`, 
we need to have another rule and the guardrail to prevent incorrect usage.

Reply:
Not against this idea, we just didn't have much feedback from the community so far, so start with a simple way sounds more charming to me. I'm ok with both paths.

And, will we consider `Wait` instead of `Pending`？

Reply:
The hints functionality is already complex. Let's ensure it's stable and uses memory efficiently, before we consider more complex scenarios.